include/pf-applications/grain_tracker/tracker.h

```diff
@@ -247,33 +241,56 @@ namespace GrainTracker
               grains_candidates.erase(new_grain_id);
             }
 
-          // Insert new grain
-          grains.emplace(std::make_pair(new_grain_id, new_grain));
-          grains.at(new_grain_id).set_grain_id(new_grain_id);
-          grains.at(new_grain_id).set_dynamics(new_dynamics);
-
-          // Update mapping if we changed the grain id
-          if (new_grain_id != current_grain_id)
+          // Insert new grain if it has been identified or new ones are allowed
+          if (new_grain_id != std::numeric_limits<unsigned int>::max())
             {
-              auto &particle_to_grain =
-                particle_ids_to_grain_ids[new_grain.get_order_parameter_id()];
+              grains.emplace(std::make_pair(new_grain_id, new_grain));
+              grains.at(new_grain_id).set_grain_id(new_grain_id);
+              grains.at(new_grain_id).set_dynamics(new_dynamics);
 
-              for (unsigned int ip = 0; ip < particle_to_grain.size(); ip++)
+              // Update mapping if we changed the grain id
+              if (new_grain_id != current_grain_id)
                 {
-                  auto &pmap = particle_to_grain[ip];
+                  auto &particle_to_grain =
+                    particle_ids_to_grain_ids[new_grain
+                                                .get_order_parameter_id()];
 
-                  if (grains_ids_changed[new_grain.get_order_parameter_id()]
-                                        [ip] == false &&
-                      pmap.first == current_grain_id)
+                  for (unsigned int ip = 0; ip < particle_to_grain.size(); ip++)
                     {
-                      pmap.first = new_grain_id;
-                      grains_ids_changed[new_grain.get_order_parameter_id()]
-                                        [ip] = true;
+                      auto &pmap = particle_to_grain[ip];
+
+                      if (grains_ids_changed[new_grain.get_order_parameter_id()]
+                                            [ip] == false &&
+                          pmap.first == current_grain_id)
+                        {
+                          pmap.first = new_grain_id;
+                          grains_ids_changed[new_grain.get_order_parameter_id()]
+                                            [ip] = true;
+                        }
                     }
                 }
             }
         }
 
+      // Throw exception if some invalid grains have been detected
+      if (!invalid_grains.empty())
+        {
+          std::ostringstream ss;
```

Comment:
Is this printed by all processes? I guess all processes have the same info so that it is enough to print it once.

Reply:
This is not printed directly here, but instead the info accumulated by `std::ostringstream` is then used as the excpetion message below:
```c++
AssertThrow(invalid_grains.empty(), ExcGrainsInconsistency(ss.str()));
```